include/swift/AST/AutoDiff.h

```diff
@@ -113,6 +113,16 @@ struct AutoDiffConfig {
   SWIFT_DEBUG_DUMP;
 };
 
+/// Key for caching SIL derivative function types.
+struct SILAutoDiffDerivativeFunctionKey {
+  SILFunctionType *originalType;
+  IndexSubset *parameterIndices;
+  IndexSubset *resultIndices;
+  AutoDiffDerivativeFunctionKind kind;
+  CanGenericSignature derivativeFnGenSig;
+  bool isReabstractionThunk;
+};
+
```

Comment:
IIUC, you don't necessarily have Types.h included here, so DenseMap can't call alignof on it. Could you include Types.h?

Reply:
Unfortunately, including `Types.h` as is isn't possible because [`Types.h` includes `AutoDiff.h`](https://github.com/apple/swift/blob/91c790662d51bf58de6210234e87ab9693a3619b/include/swift/AST/Types.h#L20).

`Types.h` uses just `AutoDiffDerivativeFunctionKind`, `DifferentiabilityKind`, and `TangentSpace` from `AutoDiff.h` - perhaps we could move those three declarations somewhere else. I'm not sure what's the best place though.

I wonder if you have suggestions on what to do?

Reply:
> IIUC, you don't necessarily have Types.h included here, so DenseMap can't call alignof on it. Could you include Types.h?

I also wonder how you discovered the issue regarding `DenseMap` and `alignof`? Did this trigger a build error somewhere?

Reply:
Issue resolved in https://github.com/apple/swift/pull/29983.